models/resnet/resnet.hpp

```diff
@@ -0,0 +1,554 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MODELS_MODELS_RESNET_RESNET_HPP
+#define MODELS_MODELS_RESNET_RESNET_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/layer/layer_types.hpp>
+#include <mlpack/methods/ann/init_rules/random_init.hpp>
+#include <mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp>
+#include <mlpack/methods/ann/init_rules/he_init.hpp>
+
+#include "./../../utils/utils.hpp"
+
+namespace mlpack {
+namespace models {
+
+/**
+ * Definition of a ResNet CNN.
+ * 
+ * @tparam OutputLayerType The output layer type used to evaluate the network.
+ * @tparam InitializationRuleType Rule used to initialize the weight matrix.
+ * @tparam ResNetVersion Version of ResNet.
+ */
+template<
+  typename OutputLayerType = ann::CrossEntropyError<>,
+  typename InitializationRuleType = ann::RandomInitialization,
+  size_t ResNetVersion = 18
+>
+class ResNet{
+ public:
+  //! Create the ResNet model.
+  ResNet();
+
+  /**
+   * ResNet constructor intializes input shape and number of classes.
+   *
+   * @param inputChannels Number of input channels of the input image.
+   * @param inputWidth Width of the input image.
+   * @param inputHeight Height of the input image.
+   * @param includeTop Must be set to true if preTrained is set to true.
+   * @param preTrained True for pre-trained weights of ImageNet,
+   *    default is false.
+   * @param numClasses Optional number of classes to classify images into,
+   *     only to be specified if includeTop is true, default is 1000.
+   */
+  ResNet(const size_t inputChannel,
+         const size_t inputWidth,
+         const size_t inputHeight,
+         const bool includeTop = true,
+         const bool preTrained = false,
+         const size_t numClasses = 1000);
+
+  /**
+   * ResNet constructor intializes input shape and number of classes.
+   *
+   * @param inputShape A three-valued tuple indicating input shape.
+   *     First value is number of channels (channels-first).
+   *     Second value is input height. Third value is input width.
+   * @param preTrained True for pre-trained weights of ImageNet,
+   *    default is false.
+   * @param numClasses Optional number of classes to classify images into,
+   *     only to be specified if includeTop is  true.
+   */
+  ResNet(std::tuple<size_t, size_t, size_t> inputShape,
+         const bool includeTop = true,
+         const bool preTrained = false,
+         const size_t numClasses = 1000);
+
+  //! Get Layers of the model.
+  ann::FFN<OutputLayerType, InitializationRuleType>& GetModel()
+  {
+    return resNet;
+  }
+
+  //! Load weights into the model and assumes the internal matrix to be
+  //  named "ResNet"
+  void LoadModel(const std::string& filePath);
```

Comment:
Do you think it makes sense to add a model name and set this to `ResNet`?

Reply:
I am not really sure what you meant by this.

Reply:
Right now we expect the user to save the model with `"ResNet"` which I think is good and should be the default, was just wondering if we should add a parameter to the Load/Save function to change the name.

Reply:
I don't think that should be needed they can already change the file name for the model to save different models and i believe it would be great to limit the internal matrix name, that way bugs or any undefined load/save behavior would be easily identifiable.